mypy/checker.py

```diff
@@ -827,15 +828,17 @@ def check_reverse_op_method(self, defn: FuncItem, typ: CallableType,
             elif isinstance(arg_type, UnionType):
                 if not arg_type.has_readable_member(other_method):
                     return
+                arg_types = list(arg_type.iter_deep())
             else:
                 return
-
-            typ2 = self.expr_checker.analyze_external_member_access(
-                other_method, arg_type, defn)
-            self.check_overlapping_op_methods(
-                typ, method, defn.info,
-                typ2, other_method, cast(Instance, arg_type),
-                defn)
+            # We check that each method is fine when dispatched on a proper self argument
+            for arg_type in arg_types:
+                typ2 = self.expr_checker.analyze_external_member_access(
+                    other_method, arg_type, defn)
+                self.check_overlapping_op_methods(
+                    typ, method, defn.info,
+                    typ2, other_method, cast(Instance, arg_type),
```

Comment:
The cast is safe (though the unsafe is not introduced by this PR). Is there an easy way to fix it? If not, can you add a TODO comment explaining that it's unsafe?